src/display/annotation_layer.js

```diff
@@ -2420,40 +2433,156 @@ class PopupElement {
-        this.#eventBus?.dispatch("switchannotationeditormode", {
-          source: this,
-          editId,
-          mode,
-          editComment: true,
-        });
+        this.#commentManager.toggleCommentPopup(this, /* isSelected = */ true);
+      },
+      { signal }
+    );
+    button.addEventListener(
+      "pointerenter",
+      () => {
+        this.#commentManager.toggleCommentPopup(
+          this,
+          /* isSelected = */ false,
+          /* visibility = */ true
+        );
+      },
+      { signal }
+    );
+    button.addEventListener(
+      "pointerleave",
+      () => {
+        this.#commentManager.toggleCommentPopup(
+          this,
+          /* isSelected = */ false,
+          /* visibility = */ false
+        );
       },
       { signal }
     );
     const { style } = button;
     style.left = `calc(${this.#commentButtonPosition[0]}%)`;
     style.top = `calc(${this.#commentButtonPosition[1]}% - var(--comment-button-dim))`;
-    if (this.#commentButtonColor) {
-      style.backgroundColor = this.#commentButtonColor;
+    if (this.commentButtonColor) {
+      style.backgroundColor = this.commentButtonColor;
     }
     parentContainer.after(button);
   }
 
+  get commentButtonColor() {
+    const {
+      data: { color, opacity },
+    } = this.#firstElement;
+    if (!color) {
+      return null;
+    }
+    return this.#parent._commentManager.makeCommentColor(color, opacity);
+  }
+
+  getData() {
+    return this.#firstElement.data;
+  }
+
+  get elementBeforePopup() {
+    return this.#commentButton;
+  }
+
+  get comment() {
+    return this.#firstElement.data.contentsObj?.str || "";
+  }
+
+  set comment(text) {
+    const element = this.#firstElement;
+    if (text) {
+      element.data.contentsObj = { str: text };
+      // TODO: Support saving the text.
+      //  element.annotationStorage.setValue(element.data.id, {
+      //    popup: { contents: text },
+      //  });
+    } else {
+      element.data.contentsObj = null;
+      element.removePopup();
+    }
+    element.data.modificationDate = new Date();
+  }
+
+  get parentBoundingClientRect() {
+    return this.#firstElement.layer.getBoundingClientRect();
+  }
+
+  setCommentButtonStates({ selected, hasPopup }) {
+    if (!this.#commentButton) {
+      return;
+    }
+    this.#commentButton.classList.toggle("selected", selected);
+    this.#commentButton.ariaExpanded = hasPopup;
+  }
+
+  setSelectedCommentButton(selected) {
+    this.#commentButton.classList.toggle("selected", selected);
+  }
+
+  get commentPopupPosition() {
+    if (this.#popupPosition) {
+      return this.#popupPosition;
+    }
+    const { x, y, height } = this.#commentButton.getBoundingClientRect();
+    const {
+      x: parentX,
+      y: parentY,
+      width: parentWidth,
+      height: parentHeight,
+    } = this.#firstElement.layer.getBoundingClientRect();
+    return [
+      (x - parentX) / parentWidth,
+      (y + height + 2 - parentY) / parentHeight,
```

Comment:
Same here as in the other PR: let's make this 2 a named constant, or add a comment what it's for.